common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java

```diff
@@ -0,0 +1,378 @@
+
+import static org.apache.arrow.c.NativeUtil.NULL;
+import static org.apache.arrow.util.Preconditions.checkState;
+
+/**
+ * Import buffers from a C Data Interface struct. We copy it from Arrow `BufferImportTypeVisitor`
+ * and fix the issue: https://github.com/apache/arrow/issues/42156.
+ */
+class CometBufferImportTypeVisitor
+    implements ArrowType.ArrowTypeVisitor<List<ArrowBuf>>, AutoCloseable {
+  private final BufferAllocator allocator;
+  private final ReferenceCountedArrowArray underlyingAllocation;
+  private final ArrowFieldNode fieldNode;
+  private final ArrowArray.Snapshot snapshot;
+  private final long[] buffers;
+  private final List<ArrowBuf> imported;
+
+  CometBufferImportTypeVisitor(
+      BufferAllocator allocator,
+      ReferenceCountedArrowArray underlyingAllocation,
+      ArrowFieldNode fieldNode,
+      ArrowArray.Snapshot snapshot,
+      long[] buffers) {
+    this.allocator = allocator;
+    this.underlyingAllocation = underlyingAllocation;
+    this.fieldNode = fieldNode;
+    this.snapshot = snapshot;
+    this.buffers = buffers;
+    this.imported = new ArrayList<>();
+  }
+
+  @Override
+  public void close() throws Exception {
+    AutoCloseables.close(imported);
+  }
+
+  @VisibleForTesting
+  ArrowBuf importBuffer(ArrowType type, int index, long capacity) {
+    return importBuffer(type, index, 0, capacity);
+  }
+
+  @VisibleForTesting
+  ArrowBuf importBuffer(ArrowType type, int index, long offset, long capacity) {
+    checkState(
+        buffers.length > index,
+        "Expected at least %s buffers for type %s, but found %s",
+        index + 1,
+        type,
+        buffers.length);
+    long bufferPtr = buffers[index] + offset;
+
+    if (bufferPtr == NULL) {
+      // C array may be NULL but only accept that if expected capacity is zero too
+      if (capacity != 0) {
+        throw new IllegalStateException(
+            String.format("Buffer %s for type %s cannot be null", index, type));
+      } else {
+        // no data in the C array, return an empty buffer
+        return allocator.getEmpty();
+      }
+    }
+
+    ArrowBuf buf = underlyingAllocation.unsafeAssociateAllocation(allocator, capacity, bufferPtr);
+    imported.add(buf);
+    return buf;
+  }
+
+  private ArrowBuf importFixedBits(ArrowType type, int index, long bitsPerSlot) {
+    final long capacity = DataSizeRoundingUtil.divideBy8Ceil(bitsPerSlot * fieldNode.getLength());
+    return importBuffer(type, index, capacity);
+  }
+
+  private ArrowBuf importFixedBytes(ArrowType type, int index, long bytesPerSlot) {
+    final long capacity = bytesPerSlot * fieldNode.getLength();
+    return importBuffer(type, index, capacity);
+  }
+
+  private ArrowBuf importOffsets(ArrowType type, long bytesPerSlot) {
+    final long capacity = bytesPerSlot * (fieldNode.getLength() + 1);
+    final long offset = snapshot.offset * bytesPerSlot;
+    return importBuffer(type, 1, offset, capacity);
+  }
```

Comment:
This class adds offset to buffer pointer. This is major change to Arrow Java `BufferImportTypeVisitor`.